File: src/class_set.f90
!=====================================================================!
! A basic implementation of an unordered set of tuples.
!=====================================================================!

module class_set

  implicit none
  
  ! The datatype.
  type :: set

     integer, allocatable :: table(:,:)
     integer              :: num_entries
     integer              :: num_tuples

   contains

     procedure :: insert
     procedure :: add_entry
     procedure :: get_entries
     procedure :: contains

     ! The destructor.
     final :: destroy

  end type set

  ! The constructor interface for a set.
  interface set
     module procedure create
  end interface set
  
contains

  !===================================================================!
  ! The constructor implementation of a set.
  !===================================================================!

  pure type(set) function create(num_tuples, max_entries) result(this)

    integer, intent(in) :: max_entries
    integer, intent(in) :: num_tuples

    this % num_tuples = num_tuples
    
    allocate(this % table(num_tuples, max_entries))
    this % table = 0
    this % num_entries = 0
    
  end function create

  !===================================================================!
  ! The destructor for a set object.
  !===================================================================!
  
  pure subroutine destroy(this)

    type(set), intent(inout) :: this

    if(allocated(this % table)) deallocate(this % table)

  end subroutine destroy

  !===================================================================!
  ! Insert a tuple into the set and report whether it was added.
  !===================================================================!

  impure type(logical) function insert(this, tuple)

    class(set), intent(inout) :: this
    integer   , intent(in)    :: tuple(:)

    ! Check whether the tuple is already in the table.
    if (this % contains(tuple) .eqv. .false.) then
       this % num_entries = this % num_entries + 1
       this % table(:, this % num_entries) = tuple(:)
       insert = .true.
    else
       insert = .false.
    end if

  end function insert

  !===================================================================!
  ! Add an entry into the set.
  !===================================================================!

  pure subroutine add_entry(this, tuple)

    class(set), intent(inout) :: this
    integer   , intent(in)    :: tuple(:)

    ! Check whether the tuple is already in the table.

    if (this % contains(tuple) .eqv. .false.) then
       this % num_entries = this % num_entries + 1
       this % table(:, this % num_entries) = tuple(:)
    end if

  end subroutine add_entry

  !===================================================================!
  ! Check whether the first argument is a subset of the second
  ! argument. (Should this move elsewhere?)
  !===================================================================!
  
  pure type(logical) function is_subset(small, big)

    integer, intent(in) :: small(:)
    integer, intent(in) :: big(:)

    integer, allocatable :: sub(:)
    integer, allocatable :: set(:)

    integer :: lensub, i

    is_subset = .false.

    if (size(small) .gt. size(big)) return

    ! Create local copies of the arrays.
    allocate(sub, source = small)
    allocate(set, source = big)

    ! Sort the two arrays.
    call isort(sub)
    call isort(set)    
    lensub = size(sub)

    !-----------------------------------------------------------------!
    ! Check whether all entries match up to the length of the
    ! smaller array.
    !-----------------------------------------------------------------!

    is_subset = .true.
    check : do i = 1, lensub
       if (any(set .eq. sub(i)) .eqv. .false.) then
          is_subset = .false.
          exit check
       end if
    end do check

    deallocate(sub,set)    

  end function is_subset
  
  !===================================================================!
  ! Sort an integer array in ascending order. (Should this move
  ! elsewhere?)
  !===================================================================!
  
  pure subroutine isort(array)

    integer, intent(inout) :: array(:)
    integer :: temp , j , k
    integer :: n

    n = ubound(array,1)

    do j = 1 , n
       do k = j + 1 , n
          if(array(j) > array(k)) then
             temp     = array(k)
             array(k) = array(j)
             array(j) = temp
          end if
       end do
    end do

  end subroutine isort

  !===================================================================!
  ! Check whether an entry is contained in the set.
  !===================================================================!

  pure type(logical) function contains(this, tuple)

    class(set), intent(in) :: this
    integer   , intent(in) :: tuple(:)
    integer :: i

    contains = .false.

    ! Walk the existing tuples and check whether the entry exists.
   check: do i = this % num_entries, 1, -1
       ! Improve this logic; it is expensive and unnecessary.
       if (is_subset(tuple, this % table(:,i)) .eqv. .true.) then
          contains = .true.
          return
       end if
    end do check

  end function contains

  !===================================================================!
  ! Get all the entries in the set as an array.
  !===================================================================!
  
  pure subroutine get_entries(this, entries)
    
    class(set), intent(in)            :: this
    integer, allocatable, intent(out) :: entries(:,:)
   
    allocate(entries(this % num_tuples, this % num_entries))
    
    entries(:,:) = this % table(:,1:this % num_entries)

  end subroutine get_entries
  
end module class_set

File: ufvm/class_set.py
#=====================================================================#
# Basic implementation of an unordered set tuple
#=====================================================================#

import numpy as np


class Set:
    """Datatype: unordered set of integer tuples.

    The Fortran `table(num_tuples, max_entries)` stores one tuple per
    column; this is preserved as a 2-D NumPy array of the same shape.
    """

    #===================================================================#
    # Constructor implementaion of set
    #
    # Fortran: interface set => create(num_tuples, max_entries)
    #===================================================================#
    def __init__(self, num_tuples, max_entries):
        self.num_tuples = num_tuples

        self.table = np.zeros((num_tuples, max_entries), dtype=int)
        self.num_entries = 0

    #===================================================================#
    # Add an entry into the set
    #===================================================================#
    def insert(self, tuple):
        # Check if tuple is in table
        if self.contains(tuple) is False:
            self.num_entries = self.num_entries + 1
            self.table[:, self.num_entries - 1] = tuple
            return True
        else:
            return False

    #===================================================================#
    # Add an entry into the set
    #===================================================================#
    def add_entry(self, tuple):
        # Check if tuple is in table
        if self.contains(tuple) is False:
            self.num_entries = self.num_entries + 1
            self.table[:, self.num_entries - 1] = tuple

    #===================================================================#
    # Check if an entry is contains in the set
    #===================================================================#
    def contains(self, tuple):
        # loop through existing tuples and find if it exists
        for i in range(self.num_entries - 1, -1, -1):
            # Improve logic. This is expensive and unnnecessary
            if is_subset(tuple, self.table[:, i]) is True:
                return True
        return False

    #===================================================================#
    # Get all the entries in the set as an array
    #===================================================================#
    def get_entries(self):
        entries = np.array(self.table[:, 0:self.num_entries])
        return entries


#===================================================================#
# Checks if the first argument is a subset of the second argument
# (move elsewhere?)
#===================================================================#
def is_subset(small, big):
    small = np.asarray(small)
    big = np.asarray(big)

    if small.size > big.size:
        return False

    # Create local copy of arrays
    sub = np.array(small)
    set_ = np.array(big)

    # Sort two arrays
    isort(sub)
    isort(set_)
    lensub = sub.size

    # Check if all entries are equal upto the length of the smallest array
    for i in range(lensub):
        if not (set_ == sub[i]).any():    # any(set .eq. sub(i)) .eqv. .false.
            return False
    return True


#===================================================================#
# Sort an integer array ! move elsewhere?
#===================================================================#
def isort(array):
    # In-place bubble sort, mirroring the Fortran (array is intent(inout)).
    n = array.shape[0]
    for j in range(n):
        for k in range(j + 1, n):
            if array[j] > array[k]:
                temp = array[k]
                array[k] = array[j]
                array[j] = temp
